### PR TITLE
SAMZA-1719: Add caching support to table-api

### PR DESCRIPTION
**Pull request by pdu-mn1**
This change adds caching support for Samza tables. This is especially
useful for remote table where the accesses can have high latency for
applications that can tolerate staleness. Caching is added in the form
of a composite table that wraps the actual table and a cache. We reuse
the ReadWriteTable interface for the cache. A simple Guava cache-based
table is provided in this change.

Original PR was inadvertently closed: https://github.com/apache/samza/pull/524

### CODE REVIEW

**Comment by pdu-mn1** [comment]
For previous review comments, please refer to: https://github.com/apache/samza/pull/524

**Comment by pdu-mn1** [comment]
Adding @weisong44 @vjagadish1989 for review

**Review by vjagadish1989 — APPROVED** [review]
Looks good to me. thank you Peng for the contribution

File: samza-core/src/main/java/org/apache/samza/table/caching/CachingTable.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+  }
+
+  @Override
+  public Map<K, V> getAll(List<K> keys) {
+    Map<K, V> getAllResult = new HashMap<>();
+    keys.stream().forEach(k -> getAllResult.put(k, get(k)));
+    return getAllResult;
+  }
+
+  @Override
```
[the remark this answers is not in the document]
Comment (pdu-mn1) [reply samza-core/src/main/java/org/apache/samza/table/caching/CachingTable.java]: I think you meant after "flush" right?

**Pull request closed without merging**